### PR TITLE
Add FPGA interchange front and backend

### PR DESCRIPTION
**Pull request by litghost**

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -99,7 +99,7 @@ foreach(item ${ARCH})
     endif()
 endforeach()
 
-set(CMAKE_CXX_STANDARD 11)
+set(CMAKE_CXX_STANDARD 14)
```
Comment (gatecat) [inline CMakeLists.txt:102]: nitpick, can you also update the reference to C++11 in the README to C++14?

File: fpga_interchange/fpga_interchange.cpp [hunk 2]
```diff
@@ -0,0 +1,826 @@
...
+    gzFile file = gzopen(filename.c_str(), "w");
+    NPNR_ASSERT(file != Z_NULL);
+
+    NPNR_ASSERT(gzwrite(file, &bytes[0], bytes.size()) == bytes.size());
+    NPNR_ASSERT(gzclose(file) == Z_OK);
+}
+
+struct StringEnumerator {
+    std::vector<std::string> strings;
+    std::map<std::string, size_t> string_to_index;
```
Comment (gatecat) [inline fpga_interchange/fpga_interchange.cpp:46]: is there a reason here for using `map` and not `unordered_map`?

Comment (litghost) [reply fpga_interchange/fpga_interchange.cpp:46]: No reason, changed.

File: CMakeLists.txt (hunk 1 above)
Comment (litghost) [reply CMakeLists.txt:102]: Fixed.

**Comment by litghost** [comment]
@gatecat I believe I've wrangled the build, and this is ready for review.

**Pull request merged**